app/src/main/java/com/example/chessclockk/vm/MainActivityVM.kt

```diff
@@ -0,0 +1,237 @@
+package com.example.chessclockk.vm
+
+import android.os.SystemClock
+import androidx.lifecycle.LiveData
+import androidx.lifecycle.MutableLiveData
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import com.example.chessclockk.convertGameAndBonusTimeToTempo
+import com.example.chessclockk.convertHHMMSSToMillis
+import com.example.chessclockk.millisToHHMMSS
+import com.example.chessclockk.usecase.TempoUseCase
+import com.example.chessclockk.vm.IMainActivityVM.MainScreenState
+import dagger.hilt.android.lifecycle.HiltViewModel
+import kotlinx.coroutines.CoroutineStart
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.isActive
+import kotlinx.coroutines.launch
+import java.util.concurrent.TimeUnit
+import javax.inject.Inject
+
+@HiltViewModel
+class MainActivityVM @Inject constructor(
+    private val tempoUseCase: TempoUseCase
+) : ViewModel(), IMainActivityVM {
+
+    private val _clockBlack = MutableLiveData<String>()
+    override val clockBlackLiveData: LiveData<String>
+        get() = _clockBlack
+
+    private val _clockWhite = MutableLiveData<String>()
+    override val clockWhiteLiveData: LiveData<String>
+        get() = _clockWhite
+
+    private val _state = MutableLiveData<MainScreenState>()
+    override val stateLiveData: LiveData<MainScreenState>
+        get() = _state
+
+    private val clockJob: Job by lazy {
+        initializeClockJob()
+    }
+
+    //TODO companion object
+    private var whiteMillisRemaining = TimeUnit.MINUTES.toMillis(3)
+    private var blackMillisRemaining = TimeUnit.MINUTES.toMillis(3)
+    private var bonusTime = TimeUnit.SECONDS.toMillis(2)
+    private var defaultFormat = "3\" 2'"
+
+    private var state = MainScreenState(
+        timeFormat = defaultFormat,
+        gameState = GameState.NEW_GAME,
+        whiteMovesCount = 0,
+        blackMovesCount = 0
+    )
+
+    private var gameState = mutableListOf<GameState>()
+    private var blackMovesCount = 0
+    private var whiteMovesCount = 0
+
+    init {
+        updateGameState(GameState.NEW_GAME)
+        updateTimeFormat()
+        initializeGame()
+    }
```

Comment:
inicjowanie VM to ciekawe zagadnienie. Nie mogę teraz znaleźć mojego ulubionego artykułu na ten temat. W skrócie przewagę ma podejście lazy. Czyli pobieramy dane jak ktoś zacznie odpytywać nasz strumień